File: ecs-trigger-lambda/handler.py
import json
import os
import logging
import boto3

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger(__name__)

ecs = boto3.client("ecs", region_name=os.environ.get("AWS_REGION", "eu-west-2"))

ECS_CLUSTER = os.environ["ECS_CLUSTER"]
ECS_TASK_DEFINITION = os.environ["ECS_TASK_DEFINITION"]
ECS_CONTAINER_NAME = os.environ["ECS_CONTAINER_NAME"]
SUBNETS = os.environ["SUBNETS"].split(",")
SECURITY_GROUPS = os.environ["SECURITY_GROUPS"].split(",")
ASSIGN_PUBLIC_IP = os.environ.get("ASSIGN_PUBLIC_IP", "ENABLED")


def lambda_handler(event, context):
    records = event.get("Records", [])
    log.info("Received %d SQS record(s)", len(records))

    results = []
    for record in records:
        try:
            body = json.loads(record["body"])
            result = trigger_ecs_task(body)
            results.append(result)
        except Exception as e:
            log.error("Failed to process record: %s | error: %s", record.get("messageId"), e)
            raise

    return {"triggered": len(results), "tasks": results}


def trigger_ecs_task(body: dict) -> dict:
    bucket = body["bucket"]
    key = body["key"]
    document_id = body["documentId"]

    log.info("Triggering ECS task for documentId=%s", document_id)

    response = ecs.run_task(
        cluster=ECS_CLUSTER,
        taskDefinition=ECS_TASK_DEFINITION,
        launchType="FARGATE",
        networkConfiguration={
            "awsvpcConfiguration": {
                "subnets": SUBNETS,
                "securityGroups": SECURITY_GROUPS,
                "assignPublicIp": ASSIGN_PUBLIC_IP,
            }
        },
        overrides={
            "containerOverrides": [
                {
                    "name": ECS_CONTAINER_NAME,
                    "environment": [
                        {"name": "BUCKET",      "value": bucket},
                        {"name": "KEY",         "value": key},
                        {"name": "DOCUMENT_ID", "value": document_id},
                    ],
                }
            ]
        },
    )

    failures = response.get("failures", [])
    if failures:
        raise RuntimeError(f"ECS RunTask failed for {document_id}: {failures}")

    task_arn = response["tasks"][0]["taskArn"]
    log.info("ECS task started: %s", task_arn)

    return {"documentId": document_id, "taskArn": task_arn}

File: workers/lambda-kickstarter/handler.js
const { ECSClient, RunTaskCommand } = require("@aws-sdk/client-ecs");

const ecs = new ECSClient({ region: process.env.AWS_REGION ?? "eu-west-2" });

exports.handler = async (event) => {
  const records = event.Records ?? [];
  console.log(`Received ${records.length} SQS record(s)`);

  const results = [];

  for (const record of records) {
    try {
      const body = JSON.parse(record.body);
      const result = await triggerEcsTask(body);
      results.push(result);
    } catch (error) {
      console.error(`Failed to process record ${record.messageId}:`, error);
      throw error;
    }
  }

  return { triggered: results.length, tasks: results };
};

async function triggerEcsTask(body) {
  const { bucket, key, documentId } = body;
  console.log(`Triggering ECS task for documentId=${documentId}`);

  const response = await ecs.send(
    new RunTaskCommand({
      cluster: process.env.ECS_CLUSTER,
      taskDefinition: process.env.ECS_TASK_DEFINITION,
      launchType: "FARGATE",
      networkConfiguration: {
        awsvpcConfiguration: {
          subnets: process.env.SUBNETS.split(","),
          securityGroups: process.env.SECURITY_GROUPS.split(","),
          assignPublicIp: process.env.ASSIGN_PUBLIC_IP ?? "ENABLED",
        },
      },
      overrides: {
        containerOverrides: [
          {
            name: process.env.ECS_CONTAINER_NAME,
            environment: [
              { name: "BUCKET", value: bucket },
              { name: "KEY", value: key },
              { name: "DOCUMENT_ID", value: documentId },
            ],
          },
        ],
      },
    }),
  );

  if (response.failures?.length) {
    throw new Error(
      `ECS RunTask failed for ${documentId}: ${JSON.stringify(response.failures)}`,
    );
  }

  const taskArn = response.tasks?.[0]?.taskArn;
  if (!taskArn) {
    throw new Error(`ECS RunTask returned no task ARN for ${documentId}`);
  }

  console.log(`ECS task started: ${taskArn}`);
  return { documentId, taskArn };
}
